nixos/flamenco, flamenco: init at 3.5

the-furry-hubofeverything opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Flamenco is a render farm management software for Blender, used in production at Blender studio.

The applications are meant to be run as portable applications, but I wanted to setup systemd services and users, and so I made a module with it.

Note for testers: By default, executable will use the directory that its running in as work directory.  

homepage: https://flamenco.blender.org/
development docs: https://flamenco.blender.org/development/getting-started/

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

the-furry-hubofeverything committed:
flamenco: init at 3.5

the-furry-hubofeverything commented:
Ofborg failure expected, waiting for merge

the-furry-hubofeverything commented:
I think I may want to split the packages into separate packages...

the-furry-hubofeverything commented:
```
node:events:495
      throw er; // Unhandled 'error' event
      ^
Error: connect EHOSTUNREACH 2606:4700::6810:1d22:443
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1555:16)
Emitted 'error' event on ClientRequest instance at:
    at TLSSocket.socketErrorListener (node:_http_client:501:9)
    at TLSSocket.emit (node:events:517:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -113,
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '2606:4700::6810:1d22',
  port: 443
}
```

Huh. This is a very specific case of "it built on my machine!"

Give me a sec to figure this out

the-furry-hubofeverything commented:
Result of `nixpkgs-review pr 260812` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flamenco</li>
  </ul>
</details>

yeah that's weird

cole-h commented:
@ofborg build flamenco flamenco.passthru.tests

robert-manchester commented:
worked for me too:
Result of `nixpkgs-review pr 260812` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flamenco</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2817

robert-manchester reviewed (approved):
LGTM

the-furry-hubofeverything commented:
One burnout later:

This still works, and I've written a module for it  - but I'd like to get this merged first

the-furry-hubofeverything commented:
The module now is fleshed out, and I've been running this on my homelab the past few days.

the-furry-hubofeverything committed:
nixos/flamenco: init

the-furry-hubofeverything commented:
Extraneous commit removed

the-furry-hubofeverything commented:
Updates:

- updated from 3.4 to 3.5 
- removed uses [`rec {...}`](https://nix.dev/guides/best-practices#recursive-attribute-set-rec) and [`with lib;`](https://nix.dev/guides/best-practices#with-attrset-expression)
- add `fixup-yarn-lock` in nativeBuildInputs

gador commented:
damn..I just stumbled over your PR. I spent some time yesterday implementing flamenco (see  https://github.com/gador/nixpkgs/tree/init-flamenco).
I had to patch the internal storage engine though (https://github.com/gador/nixpkgs/commit/21bde2b47cbb7613d49589c7293745fa8d5ca124) , because flamenco-manager always tried to store its file right next to the executable..
Did you try rendering with your pr?

the-furry-hubofeverything commented:
> damn..I just stumbled over your PR. I spent some time yesterday implementing flamenco (see https://github.com/gador/nixpkgs/tree/init-flamenco). I had to patch the internal storage engine though ([gador@21bde2b](https://github.com/gador/nixpkgs/commit/21bde2b47cbb7613d49589c7293745fa8d5ca124)) , because flamenco-manager always tried to store its file right next to the executable.. Did you try rendering with your pr?

I was actually thinking about patching the directory search reeeaaalll hard, but I decided against it because I didn't want the configuration behaviour to be different for the package. Instead I added `../../../../` to the nixos modules config, which corresponds to the package's path in the nix store. This is pretty hacky and I'm open to any suggestions.

I did make this render on my homelab! Here's the [configuration I'm using](https://github.com/the-furry-hubofeverything/hubble-systems/blob/main/hosts/common/flamenco.nix), which includes network share setup and some secrets handling, so modifications are definitely required if you want to copy it.

gador commented:
> Instead I added `../../../../` to the nixos modules config, which corresponds to the package's path in the nix store
This is actually a pretty neat workaround and not nearly as hacky as my patch :-D

I looked through your module and it looks pretty solid! Will probably test it in the future. I'd love to see this one being merged!

gador commented:
Result of `nixpkgs-review pr 260812` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>flamenco</li>
  </ul>
</details>

the-furry-hubofeverything commented:
https://github.com/NixOS/nixpkgs/assets/53921912/25a42e6c-13d7-4f95-b186-339fffcfb6e3

I made a render with flamenco 3.5 (three render workers, render manager on one of the workers) using an old project I had.

With 3.5 I could also put this behind a nginx proxy for https (keeping mind that it also uses websockets), and it works pretty well!

the-furry-hubofeverything commented:
@gador I've made a lil patch as an alternative to the local_manager_storage_path dilemma. 
https://github.com/the-furry-hubofeverything/hubble-systems/blob/c2718ae0d0f93e1cffcb8cac90c2620d77a22caf/pkgs/flamenco/absolute-path-bypass.patch#L1-L21

I thought it made intuitive sense to override the normal function of finding a relative path from the executable if and only if the path is set to an absolute path. That way, it can retain its original intended portability and I don't have to implement an workaround in the module.

I'm not going implement it yet because I want to see what opinion upstream has about it, but it's another option.
